File: Longest_Increasing_Subsequence.cpp
// Title: 300. Longest Increasing Subsequence
// Difficulty: Medium
// Problem: https://leetcode.com/problems/longest-increasing-subsequence/description/

// Using Dynamic Programming: O(n^2)
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp(nums.size(), 1);
        int max_len = 1;

        for (int i=1; i<nums.size(); i++){
            for (int j=0; j<i; j++){
                if (nums[j] < nums[i]){
                    dp[i] = max(dp[i], dp[j]+1);
                }
            }
            max_len = max(max_len, dp[i]);
        }
        return max_len;
    }
};

// Using Vectors and binary Search: O(n logn)
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
 
        vector<int> sub;

        for(int x : nums) {
            if(sub.empty() || sub[sub.size()-1] < x) {
                sub.push_back(x);
            }
            else {
                auto idx = lower_bound(sub.begin(), sub.end(), x);
                *idx = x;
            }
        }

        return sub.size();
    }
};

File: Longest_Increasing_Subsequence.py
# Title: 300. Longest Increasing Subsequence
# Difficulty: Medium
# Problem: https://leetcode.com/problems/longest-increasing-subsequence/description/

## Using Dynamic Programming: O (n^2)
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:

        dp = [1] * len(nums)
        max_len = 1

        for i in range(1, len(nums)):
            for j in range(i):
                if nums[j] < nums[i]:
                    dp[i] = max(dp[i], 1 + dp[j])
            max_len = max(max_len, dp[i])
        return max_len


## Using Binary Search and Lists: O (n logn)
class Solution:
    def lengthOfLIS(self, nums):
        sub = [nums[0]]  # Initialize an empty list to store the subsequence.

        for num in nums[1:]:
            # Find the index at which 'num' should be inserted in 'sub' using binary search.
            if num > sub[-1]:
                sub.append(num) # If 'num' should be inserted at the end of 'sub', simply append it.
            else:
                idx = bisect.bisect_left(sub, num)
                # idx = self.binarySearchInsertion(sub, num)
                sub[idx] = num  # If 'num' should replace an existing element in 'sub', update it.
        return len(sub)  # Return the length of the final LIS.
    
    '''
    Use this manual binary search, or use bisect_left from bisect library
    # bisect_left is more optimized
    
    def binarySearchInsertion(self, sub, num):
        # Manually find the index where 'num' should be inserted into 'sub' using binary search.
        left, right = 0, len(sub) - 1
        
        while left <= right:
            mid = left + (right - left) // 2  # Calculate the middle index.
            
            if sub[mid] < num:
                left = mid + 1
            else:
                right = mid - 1

        return left  # 'left' is the index where 'num' should be inserted.
    '''
